### PR TITLE
Problems utils

### PR DESCRIPTION
**Pull request by HiPeople21**
Set up Project Euler problems

### CODE REVIEW

File: sirenity/euler/__init__.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+from .problems_manager import ProblemManager
+
+__all__ = ["problem_manager"]
```
Comment (dignissimus) [inline sirenity/euler/__init__.py:3]: You don't seem to export `problem_manager`, should this be `ProblemManager`?

**Review by dignissimus — COMMENTED** [review]
Are you able to remove the `.db` files and only include the code to recreate them?

File: sirenity/euler/__init__.py (hunk 1 above)
Comment (HiPeople21) [reply sirenity/euler/__init__.py:3]: You're right. I'll change this now

**Review by dignissimus — COMMENTED** [review]
Are you able to remove the .db files and only include the scripts to recreate them?
Then we can add the db files to `.gitignore` to prevent them from being kept in git.

File: tests/test_problems.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+from sirenity.euler.problem import Problem
+
+# Clear db
+with open("test.db", "w") as f:
+    f.write("")
+
+manager = ProblemManager("test.db")
+
+
+class TestClass:
```
Comment (dignissimus) [inline tests/test_problems.py:20]:
```suggestion
class TestProblemManager:
```

The name of the test class should describe what it's testing, perhaps `TestProblemManager` as a name?

File: tests/test_problems.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+        assert len(cur.fetchall()) == 1
+        con.close()
+
+    def test_add_and_get_problem(self) -> None:
+        """Tests that problems are added and are retrieved"""
+        manager.add_to_db(prompt="Problem prompt", solution="solution")
+        problem = manager.get_at_id(1)
+        assert problem.id == 1
+        assert problem.prompt == "Problem prompt"
+        assert problem.solution == "solution"
```
Comment (dignissimus) [inline tests/test_problems.py:40]:
```suggestion
        self.assertEqual(problem.id, 1)
        self.assertEqual(problem.prompt, "Problem prompt")
        self.assertEqual(problem.solution, "solution")
```

The unittest library has functions for these

File: tests/test_problems.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        with pytest.raises(Exception):
+            manager.get_at_id(2)
+
+    def test_check_solution(self) -> None:
+        """
+        Tests that the class returns the right boolean
+        depending on if the solution is correct or not
+        """
+        assert manager.check_solution("solution", 1) == True
+        assert manager.check_solution("20933", 1) == False
```
Comment (dignissimus) [inline tests/test_problems.py:53]: Similar to the above reviews there is [`unittest.TestCase.assertTrue`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertTrue) and  [`unittest.TestCase.assertFalse`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertFalse)

**Review by dignissimus — CHANGES_REQUESTED** [review]

File: tests/test_problems.py (hunk 3 above)
Comment (HiPeople21) [reply tests/test_problems.py:40]: I though we were using pytest

File: tests/test_problems.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+    def test_create_table(self) -> None:
+        """Tests that table is created"""
+        con = sqlite3.connect(manager._file)
+        cur = con.cursor()
+        manager.create_table()
+        cur.execute(
+            "SELECT name FROM sqlite_master WHERE type='table' AND name='problems';"
+        )
+        assert len(cur.fetchall()) == 1
```
Comment (dignissimus) [inline tests/test_problems.py:31]: An `assertEqual` call would be appropriate here

File: tests/test_problems.py (hunk 3 above)
Comment (dignissimus) [reply tests/test_problems.py:40]: We are, pytest works with the unittest library

Comment (dignissimus) [reply tests/test_problems.py:40]: Oh apologies, let me rethink this

Comment (dignissimus) [reply tests/test_problems.py:40]: Yea, I'm happy using the `unittest` library right now and use pytest  to run tests and get coverage stats. But we might be able to drop pytest if something else can do coverage. We'll have to consider if I want it to keep pytest and work out what the advantages are of keeping it.

**Review by dignissimus — DISMISSED** [review]
The db files were added to the gitignore but still exist in version control. This is a funny one, you're removing them but git isn't paying attention because it's now in the gitgnore. Try git `rm --cached` on files you don't want

File: tests/test_problems.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
+import os
+import sqlite3
+import unittest
+
+from sirenity.euler import ProblemManager
+from sirenity.euler.problem import Problem
+from sirenity.euler.problems_manager import NotFoundError
+
+
+class TestProblemManager(unittest.TestCase):
```
Comment (dignissimus) [inline tests/test_problems.py:10]: Test functions should begin with `test_`

File: tests/test_problems.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    def create_table(self) -> None:
+        """Tests that table is created"""
+        con = sqlite3.connect(self.manager._file)
+        cur = con.cursor()
+        self.manager.create_table()
+        cur.execute(
+            "SELECT name FROM sqlite_master WHERE type='table' AND name='problems';"
+        )
+        self.assertEqual(len(cur.fetchall()), 1)
+        con.close()
```
Comment (dignissimus) [inline tests/test_problems.py:35]: When we write tests, we do it to test that the code we write works correctly no matter how the implementation changes. Tests shouldn't have implementation details inside them. You should be able to run `self.manager.add_to_db`.

A good example is your `add_and_get_problem` test.

File: tests/test_problems.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+            self.manager.add_to_db(prompt="Problem prompt", solution="solution")
+        self.assertTrue(isinstance(self.manager.get_random_problem(), Problem))
+
+    def test_all(self):
+        """Tests problems in order"""
+        self.create_table()
+        self.add_and_get_problem()
+        self.get_error()
+        self.check_solution()
+        self.check_random_problem()
```
Comment (dignissimus) [inline tests/test_problems.py:72]: If doing certain things are necessary to prepare for testing a function, you can [`setUp`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.setUp) and [`tearDown`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.tearDown)

File: sirenity/euler/problems_manager.py [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        dirname = os.path.dirname(__file__)
+        filename = os.path.join(dirname, "sources/prompts.txt")
+        with open(filename, "r", encoding="utf-8") as f:
+            prompts = f.read().split("==========")
+        filename = os.path.join(dirname, "sources/solutions.txt")
+        with open(filename, "r", encoding="utf-8") as f:
+            solutions = f.read().splitlines()[: len(prompts)]
+        qs = zip(prompts, solutions)
+        for index, q in enumerate(qs, 1):
+            manager.add_to_db(q[0], q[1], index)
```
Comment (dignissimus) [inline sirenity/euler/problems_manager.py:147]: The function definition can be inside the main body of the script

File: tests/test_problems.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+        self.manager.add_to_db(prompt="Problem prompt", solution="solution")
+        problem = self.manager.get_at_id(1)
+        self.assertEqual(problem.id, 1)
+        self.assertEqual(problem.prompt, "Problem prompt")
+        self.assertEqual(problem.solution, "solution")
+
+    def get_error(self) -> None:
+        """Tests that error is thrown if id is out of range"""
+        with self.assertRaises(NotFoundError):
+            self.manager.get_at_id(2)
```
Comment (dignissimus) [inline tests/test_problems.py:48]: Not asking for a change here, but this is a really good test. Additionally, I didn't even know `assertRaises` could be used like a context manager. Good work!

File: sirenity/euler/problems_manager.py (hunk 9 above)
Comment (dignissimus) [inline sirenity/euler/problems_manager.py:147]: Perhaps rename `qs` to `questions`, `f` to file and `q` to question. I didn't find it obvious what they meant

File: tests/test_problems.py [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+
+    def check_solution(self) -> None:
+        """Tests that the class returns the right boolean depending on if the solution is correct or not"""
+        if self.manager.get_number_of_problems() == 0:
```
Comment (dignissimus) [inline tests/test_problems.py:52]: General query, should this be 0 or not?

File: sirenity/euler/problems_manager.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+        """
+        Returns a random problem
+
+        :return: random problem
+        """
+        no_of_problems = self.get_number_of_problems()
+        if no_of_problems == 0:
+            raise NotFoundError("No problems found in table")
+        id = random.randint(1, self.get_number_of_problems())
+        return self.get_at_id(id)
```
Comment (dignissimus) [inline sirenity/euler/problems_manager.py:128]: `no_of_problems` can be `number_of_problems`

Comment (dignissimus) [reply sirenity/euler/problems_manager.py:128]: Would it be possible to run a query like this? `SELECT (id, prompt, solution) FROM problems ORDER BY RANDOM() LIMIT 1`? To get a random item

File: sirenity/euler/problems_manager.py [hunk 13]
```diff
@@ -0,0 +1,147 @@
+import random
+import sqlite3
+
+from .problem import Problem
+
+
+class NotFoundError(Exception):
+    """Custom Exception"""
```
Comment (dignissimus) [inline sirenity/euler/problems_manager.py:8]: As good practice, explain what the exception is used for (even though the name makes it fairly obvious 😝).
```"""Thrown when ??? is not found"""```
Maybe name it `ProblemNotFoundError`

**Review by dignissimus — COMMENTED** [review]
Overall the code was written very well and the tests were also superb. I had a few comments to do with code-style but otherwise, the quality is fantastic and I'm very pleased with this.

File: tests/test_problems.py (hunk 8 above)
Comment (dignissimus) [reply tests/test_problems.py:72]: All tests are run by default in the library. So you'd define `setUp` and `tearDown` to create the database and destroy it between test instances. If you want to setUp something only once, you can use do [`setUpClass`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.setUpClass)

File: tests/test_problems.py (hunk 11 above)
Comment (HiPeople21) [reply tests/test_problems.py:52]: When the tests were being run on Ubuntu, it seemed like the table didn't have any rows in it, even though other tests that add the row were run before it.

File: tests/test_problems.py (hunk 7 above)
Comment (HiPeople21) [reply tests/test_problems.py:35]: Sorry, I don't quite understand what you mean.

File: sirenity/euler/problems_manager.py [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+        )
+
+
+if __name__ == "__main__":
+    manager = ProblemManager()
+    manager.create_table()
+    print(manager.get_random_problem())
+    print(manager.get_random_problem(min_difficulty=10))
+    print(manager.get_random_problem(max_difficulty=50))
+    print(manager.get_random_problem(min_difficulty=50, max_difficulty=70))
```
Comment (dignissimus) [inline sirenity/euler/problems_manager.py:195]:
```suggestion
def main():
    """ Run some basic functionality tests """
    manager = ProblemManager()
    manager.create_table()
    print(manager.get_random_problem())
    print(manager.get_random_problem(min_difficulty=10))
    print(manager.get_random_problem(max_difficulty=50))
    print(manager.get_random_problem(min_difficulty=50, max_difficulty=70))

if __name__ == "__main__":
    main()
```

File: sirenity/euler/problems_manager.py [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+        row = self._cur.fetchone()
+        if row is None:
+            raise ProblemNotFoundError("No problems can be found")
+
+        return Problem(row[0], row[1], row[2], row[3])
+
+
+def add_problems():
+    """Adds problems to the database"""
+    import os
```
Comment (dignissimus) [inline sirenity/euler/problems_manager.py:152]: Even though this is only used here, it should be moved to the top of the file

**Review by dignissimus — CHANGES_REQUESTED** [review]
I'm happy merging this after these have been changed

File: sirenity/euler/problems_manager.py (hunk 15 above)
Comment (dignissimus) [reply sirenity/euler/problems_manager.py:152]: I'm referring to the import statement

File: sirenity/euler/sources/prompts.txt [hunk 16]
```diff
@@ -0,0 +1,9329 @@
+If we list all the natural numbers below 10 that are multiples of 3 or 5,
```
[the remark this answers is not in the document]
Comment (HiPeople21) [reply sirenity/euler/sources/prompts.txt]: `create_sources.py` can generate the sources files

**Review by dignissimus — APPROVED** [review]
Well done for persevering 😂

The tests are superb, extremely extensive. 

In the future, we can try storing the data as csv and try removing some of the `type: ignore`s

**Pull request merged**